lib/index.js

```diff
@@ -218,11 +225,61 @@ class Embark {
     }
 
     if (cmdPlugin) {
-      cmdPlugin.uploadCmds[0].cb();
+      async.waterfall([
+
+        function (callback){
+          engine.logger.debug('building dapp...');
+          engine.startMonitor();
+          engine.startService("libraryManager");
+          engine.startService("web3");
+          engine.startService("pipeline");
+          engine.startService("codeGenerator");
+          engine.startService("deployment");
+          engine.startService("ipfs");
+          
+          // 3. upload to storage (outputDone event triggered after webpack finished)
+          engine.events.on('outputDone', function () {
+            engine.logger.debug('deploying to ' + platform + '...');
+            cmdPlugin.uploadCmds[0].cb()
+            .then((success) => {
+              callback(null, success);
+            })
+            .catch((err) => {
+              callback(err);
+            });
+            
+          });
+
+          // 1. build the contracts
+          engine.deployManager.deployContracts(function (err) {
+            if(err){
+              callback(err);
```

Comment:
should this be return? or do we want the following step triggered regardless?

Reply:
If there is an `err` present in the callback of the `deployContracts` method, shouldn't the waterfall function exit and hit the callback passed in to the `async.waterfall` instead of executing the following step (which is `engine.events.emit('asset-changed', engine.contractsManager);`)? Even though it's an `async.waterfall`, there really is only one step, so as to utilise error handling and messaging.

Reply:
since it's async, it wouldn't exist immediatly and still return the code following it, the only way to avoid this to return immediatly

Reply:
Have removed this code in the latest updates.